app/src/kscan_sideband_behaviors.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2023 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_kscan_sideband_behaviors
+
+#include <zephyr/device.h>
+#include <drivers/behavior.h>
+#include <zephyr/drivers/kscan.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/device.h>
+
+#include <zmk/event_manager.h>
+#include <zmk/behavior.h>
+#include <zmk/keymap.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+struct ksbb_entry {
+    struct zmk_behavior_binding binding;
+    uint8_t row;
+    uint8_t column;
+};
+
+struct ksbb_config {
+    const struct device *kscan;
+    struct ksbb_entry *entries;
+    size_t entries_len;
+};
+
+struct ksbb_data {
+    kscan_callback_t callback;
+    bool enabled;
+};
+
+#define GET_KSBB_DEV(n) DEVICE_DT_GET(DT_DRV_INST(n)),
+
+// The kscan callback has no context with it, so we keep a static array of all possible
+// KSBBs to check when a kscan callback from the "wrapped" inner kscan fires.
+static const struct device *ksbbs[] = {DT_INST_FOREACH_STATUS_OKAY(GET_KSBB_DEV)};
+
+const struct device *find_ksbb_for_inner(const struct device *inner_dev) {
+    for (int i = 0; i < ARRAY_SIZE(ksbbs); i++) {
+        const struct device *ksbb = ksbbs[i];
+        const struct ksbb_config *cfg = ksbb->config;
+
+        if (cfg->kscan == inner_dev) {
+            return ksbb;
+        }
+    }
+
+    return NULL;
+}
+
+struct ksbb_entry *find_sideband_behavior(const struct device *dev, uint32_t row, uint32_t column) {
+    const struct ksbb_config *cfg = dev->config;
+
+    for (int e = 0; e < cfg->entries_len; e++) {
+        struct ksbb_entry *candidate = &cfg->entries[e];
+
+        if (candidate->row == row && candidate->column == column) {
+            return candidate;
+        }
+    }
+
+    return NULL;
+}
+
+void ksbb_inner_kscan_callback(const struct device *dev, uint32_t row, uint32_t column,
+                               bool pressed) {
+    struct ksbb_entry *entry = NULL;
```

Comment:
Can this be const too?

Reply:
Sadly no, that leads the inner fields to be considered `const`, which fails when calling `behavior_keymap_binding_pressed`